skada/deep/utils.py

```diff
@@ -52,3 +54,165 @@ def check_generator(seed):
+        for a single run.
+
+    tol : float, default=1e-4
+        Relative tolerance with regards to Frobenius norm of the difference
+        in the cluster centers of two consecutive iterations to declare
+        convergence.
+
+    centroids : torch.Tensor or None, default=None
+        Initial centroids to use. If None, centroids are initialized randomly.
+
+    random_state : int or None, default=None
+        Determines random number generation for centroid initialization.
+
+    device : str or torch.device, default='cpu'
+        The device to use for computations.
+
+    Attributes
+    ----------
+    cluster_centers_ : torch.Tensor of shape (n_clusters, n_features)
+        Coordinates of cluster centers.
+
+    labels_ : torch.Tensor of shape (n_samples,)
+        Labels of each point.
+
+    inertia_ : float
+        Sum of squared distances of samples to their closest cluster center.
+
+    n_iter_ : int
+        Number of iterations run.
+
+    References
+    ----------
+    Hornik, K., Feinerer, I., Kober, M., & Buchta, C. (2012). Spherical k-Means Clustering. 
+    Journal of Statistical Software, 50(10), 1–22. https://doi.org/10.18637/jss.v050.i10
+    """
+
+    def __init__(self, n_clusters=8, n_init=10, max_iter=300, tol=1e-4, 
+                 centroids=None, random_state=None, device='cpu'):
+        self.n_clusters = n_clusters
+        self.n_init = n_init
+        self.max_iter = max_iter
+        self.tol = tol
+        self.centroids = centroids
+        self.random_state = random_state
+        self.device = device
+
+    def _init_centroids(self, X):
+        # Randomly initialize centroids
+        n_samples = X.shape[0]
+        indices = torch.randperm(n_samples, device=self.device)[:self.n_clusters]
+        centroids = X[indices]
+        return centroids / torch.norm(centroids, dim=1, keepdim=True)
+
+    def fit(self, X, y=None):
+        """Compute spherical k-means clustering.
+
+        Parameters
+        ----------
+        X : torch.Tensor of shape (n_samples, n_features)
+            Training instances to cluster.
+
+        y : Ignored
+            Not used, present here for API consistency by convention.
+
+        Returns
+        -------
+        self : object
+            Fitted estimator.
+        """
+        if not isinstance(X, torch.Tensor):
+            X = torch.tensor(X, dtype=torch.float32, device=self.device)
+        else:
+            X = X.to(self.device)
+
+        best_inertia = None
+        best_centroids = None
+        best_n_iter = None
+
+        for _ in range(self.n_init):
+            if self.centroids is None:
+                centroids = self._init_centroids(X)
+            else:
+                centroids = self.centroids.to(self.device)
+
+            for n_iter in range(self.max_iter):
+                # Assign samples to closest centroids
+                dissimilarities = self._compute_dissimilarities(X, centroids)
```

Comment:
There is a difference here with the function `cosine_similarities` de torch?

Reply:
In paper: `cosine_dissimilarity` is  `0.5*(1 − cosine_similarity)`